File: string/487-lengthOfLongestSubstring.py

def longestSubstring(s):
    n=len(s)
    L=R=0
    ans=0
    win=set()
    d=dict()
    while R<n:
        x=s[R]
        i=d.get(x,-1)+1
        while L<i:
            win.remove(s[L])
            L+=1
        d[x]=R
        win.add(s[R])
        R+=1
        ans=max(ans,len(win))
    return ans

s="abcbbc"
print(longestSubstring(s))

File: string/487-lengthOfLongestSubstring.cpp
#include<iostream>
#include<map>
#include<string>
#include<set>

using STR=std::string;
using SET=std::set<char>;
using MAP=std::map<char,int>;

int longestSubString(const STR& s )
{
    int n=s.size();
    MAP d{};
    int r=0,ans=0,l=0;
    SET st{};
    while(r<n)
    {
        char x=s[r];
        int idx=d[x];
        while(l<idx)
        {
            st.erase(s[l]);
            l+=1;
        }
        st.insert(x);
        ans=std::max(ans,static_cast<int>(st.size()));
        d[x]=r;
        r+=1;
    }
    return ans;
}
int main()
{
    STR s{"abcdebbc"};
    std::cout<<longestSubString(s);
}